runtime/src/main/java/org/corfudb/runtime/collections/DiskBackedCorfuTable.java

```diff
@@ -0,0 +1,491 @@
+                            indexToId.get(index.getName().get()), secondaryKey, primaryKey);
+
+                    // We need to persist the actual value size, since multi-get API
+                    // requires an allocation of a direct buffer.
+                    final ByteBuf serializedIndexValue = Unpooled.buffer();
+                    serializedIndexValue.writeInt(valueSize);
+
+                    try {
+                        rocksApi.insert(columnFamilyRegistry.getSecondaryIndexColumnFamily(),
+                                serializedCompoundKey, serializedIndexValue);
+                    } finally {
+                        serializedCompoundKey.release();
+                        serializedIndexValue.release();
+                    }
+                }
+            }
+        } catch (Exception fatal) {
+            log.error("Received an exception while computing the index. " +
+                    "This is most likely an issue with the client's indexing function.", fatal);
+
+            close(); // Do not leave the table in an inconsistent state.
+
+            // In case of both a transactional and non-transactional operation, the client
+            // is going to receive UnrecoverableCorfuError along with the appropriate cause.
+            throw new UnrecoverableCorfuError(fatal);
+        }
+    }
+
+    public DiskBackedCorfuTable<K, V> clear() {
+        try {
+            rocksApi.clear();
+        } catch (RocksDBException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        return this;
+    }
+
+    public Stream<Map.Entry<K, V>> entryStream() {
+        final RocksDbEntryIterator<K, V> entryIterator = rocksApi.getIterator(serializer);
+        Stream<Map.Entry<K, V>> resStream = StreamSupport.stream(
+                Spliterators.spliteratorUnknownSize(entryIterator, Spliterator.ORDERED), false);
+        return resStream.onClose(entryIterator::close);
+    }
+
+    public long size() {
+        return rocksApi.exactSize();
+    }
+
+    public <I> Iterable<Map.Entry<K, V>> getByIndex(@NonNull final Index.Name indexName,
+                                                    @NonNull I indexKey) {
+        final List<ByteBuffer> keys = new ArrayList<>();
+        final List<ByteBuffer> values = new ArrayList<>();
+
+        try {
+            String secondaryIndex = indexName.get();
+            if (!secondaryIndexesAliasToPath.containsKey(secondaryIndex)) {
+                return null;
+            }
+
+            byte indexId = indexToId.get(secondaryIndexesAliasToPath.get(secondaryIndex));
+            rocksApi.prefixScan(columnFamilyRegistry.getSecondaryIndexColumnFamily(),
+                    indexId, indexKey, serializer, keys, values);
+
+            // Prevent the keys from being consumed.
+            final List<ByteBuffer> duplicateKeys = keys.stream().map(ByteBuffer::duplicate)
+                    .collect(Collectors.toList());
```

Comment:
We really need _all_ primary keys collected before we issue a `multiGet`, which relies on batching to amortize the cost. Otherwise, we would be forced to perform sequential `get`, which is painfully slow.

You also might be wondering regarding `ByteBuffer::duplicate`. We are actually not duplicating the content of the buffer, but only the pointers -- capacity, limit, position, and mark. Here is the JavaDoc:

    /**
     * Creates a new byte buffer that shares this buffer's content.
     *
     * <p> The content of the new buffer will be that of this buffer.  Changes
     * to this buffer's content will be visible in the new buffer, and vice
     * versa; the two buffers' position, limit, and mark values will be
     * independent.
     *
     * <p> The new buffer's capacity, limit, position, and mark values will be
     * identical to those of this buffer.  The new buffer will be direct if,
     * and only if, this buffer is direct, and it will be read-only if, and
     * only if, this buffer is read-only.  </p>
     *
     * @return  The new byte buffer
     */

Reply:
It's not surprising that `ByteBuffer::duplicate` shares the underlying data content which was what I assumed. My point is that this getByIndex() function might need lots of memory space to store the result (`key` + `values`). Since getByIndex() returns an iterator, can we not run multiGet at once but instead issue multiple multiGet as the iterator advances?

Reply:
The API `getByIndex` returns a collection, therefore, all of the data has to be loaded into memory anyways. If the API was to change to return a stream, would no longer be moot. I am also guessing that the clients would just go ahead and collect the result anyways.